### PR TITLE
16.0 normalization woes

### PR DESCRIPTION
**Pull request by eggrobin**
See unicode-org/properties#206.

This PR fixes the derivation of the NF<i>meow</i>C_QC properties, and adds:
- A test that the quickCheck algorithm defined in UAX15 is correct on all columns of all test cases in NormalizationTest.txt, for both normalization forms (failed thanks to the new parts before fixing the derivations; the existing parts would not have caught it).
- An invariant test that NF<i>meow</i>C_QC≠Yes on characters whose _meow_ decomposition starts with a character with NF<i>meow</i>C_QC≠Yes (failed before fixing the derivations).
- Part 4 of NormalizationTest.txt « canonical closures (excluding Hangul) », all strings canonically equivalent to single code points, excluding NFD and NFC (already covered in Part 1) and excluding Hangul.
- Part 5 of NormalizationTest.txt « chained primary composites ». All chains of primary composites, defined below, appear in this part. Most of them appear in c1, but some appear only in c2 (NFC).
   - Let X and Y be primary composites or non-decomposable starters.
   - Let S be canonically equivalent to the concatenation of X and Y.
   - S is a chain of primary composites if both of the following hold:
      1. S does not split into two substrings canonically equivalent to X and Y;
      2. S is not canonically equivalent to a single code point.
         - Note: If S is canonically equivalent to a single code point, it is covered by the earlier parts.
         - Note: This implies that at least one of X and Y is a primary composite.

Test failures before fixing the derivations:
```
[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestProperties>TestFmwkMinusMinus.tearDown:23 errln()
NFC quickCheck returns YES for non-normalized string KIRAT RAI VOWEL SIGN AA, KIRAT RAI VOWEL SIGN AI (@Part4 # Canonical closures (excluding Hangul))
NFC quickCheck returns YES for non-normalized string TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN OO (@Part5 # Chained primary composites)
NFC quickCheck returns YES for non-normalized string TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AI (@Part5 # Chained primary composites)
NFC quickCheck returns YES for non-normalized string TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AU (@Part5 # Chained primary composites)
NFC quickCheck returns YES for non-normalized string KIRAT RAI VOWEL SIGN O, KIRAT RAI VOWEL SIGN AI (@Part5 # Chained primary composites) ==> expected: <0> but was: <5>
[ERROR]   TestProperties>TestFmwkMinusMinus.tearDown:23 errln()
U+113C5 (𑏅) TULU-TIGALARI VOWEL SIGN AI has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+113C2 (𑏂) TULU-TIGALARI VOWEL SIGN EE, which has NFC_QC=Maybe
U+113C7 (𑏇) TULU-TIGALARI VOWEL SIGN OO has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+113C2 (𑏂) TULU-TIGALARI VOWEL SIGN EE, which has NFC_QC=Maybe
U+113C8 (𑏈) TULU-TIGALARI VOWEL SIGN AU has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+113C2 (𑏂) TULU-TIGALARI VOWEL SIGN EE, which has NFC_QC=Maybe
U+16121 (𖄡) GURUNG KHEMA VOWEL SIGN U has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFC_QC=Maybe
U+16122 (𖄢) GURUNG KHEMA VOWEL SIGN UU has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFC_QC=Maybe
U+16123 (𖄣) GURUNG KHEMA VOWEL SIGN E has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFC_QC=Maybe
U+16124 (𖄤) GURUNG KHEMA VOWEL SIGN EE has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+16129 (𖄩) GURUNG KHEMA VOWEL LENGTH MARK, which has NFC_QC=Maybe
U+16125 (𖄥) GURUNG KHEMA VOWEL SIGN AI has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFC_QC=Maybe
U+16D68 (𖵨) KIRAT RAI VOWEL SIGN AI has NFC_QC=Yes, but its (canonical) Decomposition_Mapping starts with U+16D67 (𖵧) KIRAT RAI VOWEL SIGN E, which has NFC_QC=Maybe ==> expected: <0> but was: <9>
[ERROR]   TestProperties>TestFmwkMinusMinus.tearDown:23 errln()
NFKC quickCheck returns YES for non-normalized string KIRAT RAI VOWEL SIGN AA, KIRAT RAI VOWEL SIGN AI (@Part4 # Canonical closures (excluding Hangul))
NFKC quickCheck returns YES for non-normalized string TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN OO (@Part5 # Chained primary composites)
NFKC quickCheck returns YES for non-normalized string TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AI (@Part5 # Chained primary composites)
NFKC quickCheck returns YES for non-normalized string TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AU (@Part5 # Chained primary composites)
NFKC quickCheck returns YES for non-normalized string KIRAT RAI VOWEL SIGN O, KIRAT RAI VOWEL SIGN AI (@Part5 # Chained primary composites) ==> expected: <0> but was: <5>
[ERROR]   TestProperties>TestFmwkMinusMinus.tearDown:23 errln()
U+113C5 (𑏅) TULU-TIGALARI VOWEL SIGN AI has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+113C2 (𑏂) TULU-TIGALARI VOWEL SIGN EE, which has NFKC_QC=Maybe
U+113C7 (𑏇) TULU-TIGALARI VOWEL SIGN OO has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+113C2 (𑏂) TULU-TIGALARI VOWEL SIGN EE, which has NFKC_QC=Maybe
U+113C8 (𑏈) TULU-TIGALARI VOWEL SIGN AU has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+113C2 (𑏂) TULU-TIGALARI VOWEL SIGN EE, which has NFKC_QC=Maybe
U+16121 (𖄡) GURUNG KHEMA VOWEL SIGN U has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFKC_QC=Maybe
U+16122 (𖄢) GURUNG KHEMA VOWEL SIGN UU has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFKC_QC=Maybe
U+16123 (𖄣) GURUNG KHEMA VOWEL SIGN E has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFKC_QC=Maybe
U+16124 (𖄤) GURUNG KHEMA VOWEL SIGN EE has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+16129 (𖄩) GURUNG KHEMA VOWEL LENGTH MARK, which has NFKC_QC=Maybe
U+16125 (𖄥) GURUNG KHEMA VOWEL SIGN AI has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+1611E (𖄞) GURUNG KHEMA VOWEL SIGN AA, which has NFKC_QC=Maybe
U+16D68 (𖵨) KIRAT RAI VOWEL SIGN AI has NFKC_QC=Yes, but its Decomposition_Mapping starts with U+16D67 (𖵧) KIRAT RAI VOWEL SIGN E, which has NFKC_QC=Maybe ==> expected: <0> but was: <9>
[INFO]
[ERROR] Tests run: 15, Failures: 4, Errors: 0, Skipped: 1
```

### CODE REVIEW

**Comment by markusicu** [comment]
(I edited/fixed my response.)

**Comment by eggrobin** [comment]
Since we are going to want good conformance tests for 16.0 α, I (very stupidly) generated NormalizationTest.txt test cases for the closure under canonical equivalence of every canonical decomposition. It’s mostly Hangul by volume, we easily could drop that part if we wanted.

The following test cases seem interesting, so far we had only looked at Kirat Rai and Tulu-Tigalari, but it looks like Gurung Khema does the multi-path thing too:
```
1611E 16123;16126;1611E 1611E 1611F;16126;1611E 1611E 1611F; # (◌𖄞◌𖄣; ◌𖄦; ◌𖄞◌𖄞◌𖄟; ◌𖄦; ◌𖄞◌𖄞◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN E
16121 1611F;16126;1611E 1611E 1611F;16126;1611E 1611E 1611F; # (◌𖄡◌𖄟; ◌𖄦; ◌𖄞◌𖄞◌𖄟; ◌𖄦; ◌𖄞◌𖄞◌𖄟; ) GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN I
1611E 16124;16127;1611E 16129 1611F;16127;1611E 16129 1611F; # (◌𖄞◌𖄤; ◌𖄧; ◌𖄞◌𖄩◌𖄟; ◌𖄧; ◌𖄞◌𖄩◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN EE
16122 1611F;16127;1611E 16129 1611F;16127;1611E 16129 1611F; # (◌𖄢◌𖄟; ◌𖄧; ◌𖄞◌𖄩◌𖄟; ◌𖄧; ◌𖄞◌𖄩◌𖄟; ) GURUNG KHEMA VOWEL SIGN UU, GURUNG KHEMA VOWEL SIGN I
1611E 16125;16128;1611E 1611E 16120;16128;1611E 1611E 16120; # (◌𖄞◌𖄥; ◌𖄨; ◌𖄞◌𖄞◌𖄠; ◌𖄨; ◌𖄞◌𖄞◌𖄠; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN AI
16121 16120;16128;1611E 1611E 16120;16128;1611E 1611E 16120; # (◌𖄡◌𖄠; ◌𖄨; ◌𖄞◌𖄞◌𖄠; ◌𖄨; ◌𖄞◌𖄞◌𖄠; ) GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN II
16D63 16D68;16D6A;16D63 16D67 16D67;16D6A;16D63 16D67 16D67; # (𖵪; 𖵪; 𖵪; 𖵪; 𖵪; ) KIRAT RAI VOWEL SIGN AA, KIRAT RAI VOWEL SIGN AI
16D69 16D67;16D6A;16D63 16D67 16D67;16D6A;16D63 16D67 16D67; # (𖵪; 𖵪; 𖵪; 𖵪; 𖵪; ) KIRAT RAI VOWEL SIGN O, KIRAT RAI VOWEL SIGN E
```

Additional test cases should be added to properly cover the spukhafte Fernwirkung from the two E E = AI decompositions; that only gets really interesting when considering sequences that are more than one character long in NFC.

**Comment by markusicu** [comment]
> Since we are going to want good conformance tests for 16.0 α, I (very stupidly) generated NormalizationTest.txt test cases for the closure under canonical equivalence of every canonical decomposition.

Thanks!

> It’s mostly Hangul by volume, we easily could drop that part if we wanted.

Yes, let's drop Hangul.
The file is too large to view in the browser, but that data is likely boring. Hangul normalization is done in code and easily tested with just a handful of cases.

> The following test cases seem interesting, so far we had only looked at Kirat Rai and Tulu-Tigalari, but it looks like Gurung Khema does the multi-path thing too:

I see... U=AA+AA, and several other vowel signs have decomps starting with AA. (U itself, UU, E, AI, O, OO, AU)

> Additional test cases should be added to properly cover the spukhafte Fernwirkung from the two E E = AI decompositions; that only gets really interesting when considering sequences that are more than one character long in NFC.

Yes. In particular, we should hardcode some additional test cases like Kirat Rai AA+AI+E, AA+E+AI, E+AI+AI, O+AI, O+AI+AI.

**Comment by markusicu** [comment]
Gurung Khema has another overlap: UU=AA+length, and EE=length+I --> AA+EE=UU+I

I added Gurung Khema to the title and writeup of https://github.com/unicode-org/properties/issues/206

**Comment by eggrobin** [comment]
> Yes, let's drop Hangul.

Done.

> > Additional test cases should be added to properly cover the spukhafte Fernwirkung from the two E E = AI decompositions; that only gets really interesting when considering sequences that are more than one character long in NFC.
>
> Yes. In particular, we should hardcode some additional test cases like Kirat Rai AA+AI+E, AA+E+AI, E+AI+AI, O+AI, O+AI+AI.

I tried generating that instead of hardcoding it, which found Gurung Khema:
```
1138B 113C5 113C2;1138E 113C5;1138B 113C2 113C2 113C2;1138E 113C5;1138B 113C2 113C2 113C2; # (𑎎𑏅; 𑎎𑏅; 𑎎𑏅; 𑎎𑏅; 𑎎𑏅; ) TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AI, TULU-TIGALARI VOWEL SIGN EE
1138B 113C5 113B8;1138E 113C7;1138B 113C2 113C2 113B8;1138E 113C7;1138B 113C2 113C2 113B8; # (𑎎𑏇; 𑎎𑏇; 𑎎𑏇; 𑎎𑏇; 𑎎𑏇; ) TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AI, TULU-TIGALARI VOWEL SIGN AA
1138B 113C5 113C9;1138E 113C8;1138B 113C2 113C2 113C9;1138E 113C8;1138B 113C2 113C2 113C9; # (𑎎𑏈; 𑎎𑏈; 𑎎𑏈; 𑎎𑏈; 𑎎𑏈; ) TULU-TIGALARI LETTER EE, TULU-TIGALARI VOWEL SIGN AI, TULU-TIGALARI AU LENGTH MARK
113C2 113C5 113C2;113C5 113C5;113C2 113C2 113C2 113C2;113C5 113C5;113C2 113C2 113C2 113C2; # (𑏅𑏅; 𑏅𑏅; 𑏅𑏅; 𑏅𑏅; 𑏅𑏅; ) TULU-TIGALARI VOWEL SIGN EE, TULU-TIGALARI VOWEL SIGN AI, TULU-TIGALARI VOWEL SIGN EE
113C2 113C5 113B8;113C5 113C7;113C2 113C2 113C2 113B8;113C5 113C7;113C2 113C2 113C2 113B8; # (𑏅𑏇; 𑏅𑏇; 𑏅𑏇; 𑏅𑏇; 𑏅𑏇; ) TULU-TIGALARI VOWEL SIGN EE, TULU-TIGALARI VOWEL SIGN AI, TULU-TIGALARI VOWEL SIGN AA
113C2 113C5 113C9;113C5 113C8;113C2 113C2 113C2 113C9;113C5 113C8;113C2 113C2 113C2 113C9; # (𑏅𑏈; 𑏅𑏈; 𑏅𑏈; 𑏅𑏈; 𑏅𑏈; ) TULU-TIGALARI VOWEL SIGN EE, TULU-TIGALARI VOWEL SIGN AI, TULU-TIGALARI AU LENGTH MARK
1611E 16121 1611E;16121 16121;1611E 1611E 1611E 1611E;16121 16121;1611E 1611E 1611E 1611E; # (◌𖄞◌𖄡◌𖄞; ◌𖄡◌𖄡; ◌𖄞◌𖄞◌𖄞◌𖄞; ◌𖄡◌𖄡; ◌𖄞◌𖄞◌𖄞◌𖄞; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN AA
1611E 16121 16129;16121 16122;1611E 1611E 1611E 16129;16121 16122;1611E 1611E 1611E 16129; # (◌𖄞◌𖄡◌𖄩; ◌𖄡◌𖄢; ◌𖄞◌𖄞◌𖄞◌𖄩; ◌𖄡◌𖄢; ◌𖄞◌𖄞◌𖄞◌𖄩; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL LENGTH MARK
1611E 16126;16121 16123;1611E 1611E 1611E 1611F;16121 16123;1611E 1611E 1611E 1611F; # (◌𖄞◌𖄦; ◌𖄡◌𖄣; ◌𖄞◌𖄞◌𖄞◌𖄟; ◌𖄡◌𖄣; ◌𖄞◌𖄞◌𖄞◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN O
1611E 16121 1611F;16121 16123;1611E 1611E 1611E 1611F;16121 16123;1611E 1611E 1611E 1611F; # (◌𖄞◌𖄡◌𖄟; ◌𖄡◌𖄣; ◌𖄞◌𖄞◌𖄞◌𖄟; ◌𖄡◌𖄣; ◌𖄞◌𖄞◌𖄞◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN I
1611E 16128;16121 16125;1611E 1611E 1611E 16120;16121 16125;1611E 1611E 1611E 16120; # (◌𖄞◌𖄨; ◌𖄡◌𖄥; ◌𖄞◌𖄞◌𖄞◌𖄠; ◌𖄡◌𖄥; ◌𖄞◌𖄞◌𖄞◌𖄠; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN AU
1611E 16121 16120;16121 16125;1611E 1611E 1611E 16120;16121 16125;1611E 1611E 1611E 16120; # (◌𖄞◌𖄡◌𖄠; ◌𖄡◌𖄥; ◌𖄞◌𖄞◌𖄞◌𖄠; ◌𖄡◌𖄥; ◌𖄞◌𖄞◌𖄞◌𖄠; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN II
1611E 16121 16123;16121 16126;1611E 1611E 1611E 1611E 1611F;16121 16126;1611E 1611E 1611E 1611E 1611F; # (◌𖄞◌𖄡◌𖄣; ◌𖄡◌𖄦; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄟; ◌𖄡◌𖄦; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN E
1611E 16121 1611E 1611F;16121 16126;1611E 1611E 1611E 1611E 1611F;16121 16126;1611E 1611E 1611E 1611E 1611F; # (◌𖄞◌𖄡◌𖄞◌𖄟; ◌𖄡◌𖄦; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄟; ◌𖄡◌𖄦; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN I
1611E 16121 16124;16121 16127;1611E 1611E 1611E 16129 1611F;16121 16127;1611E 1611E 1611E 16129 1611F; # (◌𖄞◌𖄡◌𖄤; ◌𖄡◌𖄧; ◌𖄞◌𖄞◌𖄞◌𖄩◌𖄟; ◌𖄡◌𖄧; ◌𖄞◌𖄞◌𖄞◌𖄩◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN EE
1611E 16121 16129 1611F;16121 16127;1611E 1611E 1611E 16129 1611F;16121 16127;1611E 1611E 1611E 16129 1611F; # (◌𖄞◌𖄡◌𖄩◌𖄟; ◌𖄡◌𖄧; ◌𖄞◌𖄞◌𖄞◌𖄩◌𖄟; ◌𖄡◌𖄧; ◌𖄞◌𖄞◌𖄞◌𖄩◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL LENGTH MARK, GURUNG KHEMA VOWEL SIGN I
1611E 16121 16125;16121 16128;1611E 1611E 1611E 1611E 16120;16121 16128;1611E 1611E 1611E 1611E 16120; # (◌𖄞◌𖄡◌𖄥; ◌𖄡◌𖄨; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄠; ◌𖄡◌𖄨; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄠; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN AI
1611E 16121 1611E 16120;16121 16128;1611E 1611E 1611E 1611E 16120;16121 16128;1611E 1611E 1611E 1611E 16120; # (◌𖄞◌𖄡◌𖄞◌𖄠; ◌𖄡◌𖄨; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄠; ◌𖄡◌𖄨; ◌𖄞◌𖄞◌𖄞◌𖄞◌𖄠; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN U, GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN II
1611E 16127;16121 16124;1611E 1611E 16129 1611F;16121 16124;1611E 1611E 16129 1611F; # (◌𖄞◌𖄧; ◌𖄡◌𖄤; ◌𖄞◌𖄞◌𖄩◌𖄟; ◌𖄡◌𖄤; ◌𖄞◌𖄞◌𖄩◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN OO
1611E 16122 1611F;16121 16124;1611E 1611E 16129 1611F;16121 16124;1611E 1611E 16129 1611F; # (◌𖄞◌𖄢◌𖄟; ◌𖄡◌𖄤; ◌𖄞◌𖄞◌𖄩◌𖄟; ◌𖄡◌𖄤; ◌𖄞◌𖄞◌𖄩◌𖄟; ) GURUNG KHEMA VOWEL SIGN AA, GURUNG KHEMA VOWEL SIGN UU, GURUNG KHEMA VOWEL SIGN I
16D67 16D68 16D67;16D68 16D68;16D67 16D67 16D67 16D67;16D68 16D68;16D67 16D67 16D67 16D67; # (𖵨𖵨; 𖵨𖵨; 𖵨𖵨; 𖵨𖵨; 𖵨𖵨; ) KIRAT RAI VOWEL SIGN E, KIRAT RAI VOWEL SIGN AI, KIRAT RAI VOWEL SIGN E
16D6A 16D67;16D6A 16D67;16D63 16D67 16D67 16D67;16D6A 16D67;16D63 16D67 16D67 16D67; # (𖵪𖵧; 𖵪𖵧; 𖵪𖵧; 𖵪𖵧; 𖵪𖵧; ) KIRAT RAI VOWEL SIGN AU, KIRAT RAI VOWEL SIGN E
16D63 16D68 16D67;16D6A 16D67;16D63 16D67 16D67 16D67;16D6A 16D67;16D63 16D67 16D67 16D67; # (𖵪𖵧; 𖵪𖵧; 𖵪𖵧; 𖵪𖵧; 𖵪𖵧; ) KIRAT RAI VOWEL SIGN AA, KIRAT RAI VOWEL SIGN AI, KIRAT RAI VOWEL SIGN E
16D69 16D68 16D67;16D6A 16D68;16D63 16D67 16D67 16D67 16D67;16D6A 16D68;16D63 16D67 16D67 16D67 16D67; # (𖵪𖵨; 𖵪𖵨; 𖵪𖵨; 𖵪𖵨; 𖵪𖵨; ) KIRAT RAI VOWEL SIGN O, KIRAT RAI VOWEL SIGN AI, KIRAT RAI VOWEL SIGN E
16D63 16D67 16D68 16D67;16D6A 16D68;16D63 16D67 16D67 16D67 16D67;16D6A 16D68;16D63 16D67 16D67 16D67 16D67; # (𖵪𖵨; 𖵪𖵨; 𖵪𖵨; 𖵪𖵨; 𖵪𖵨; ) KIRAT RAI VOWEL SIGN AA, KIRAT RAI VOWEL SIGN E, KIRAT RAI VOWEL SIGN AI, KIRAT RAI VOWEL SIGN E
```

**Comment by markusicu** [comment]
Very good! Thank you!!

**Comment by eggrobin** [comment]
(The generator for the chained decompositions also picks up some weird old stuff that I didn’t mean to pick up, but I’ll try to figure that out next year.)

**Comment by eggrobin** [comment]
@markusicu I still need to fix the code that generates the NF<i>meow</i>C_QC generator so it passes the new test, but the changes to GenerateData.java (to generate the new parts of NormalizationTest.txt) should be reviewable.

**Comment by eggrobin** [comment]
@markusicu 
> And most of [the Gurung Khema] vowel signs are NFC_QC=Yes but need to be Maybe.

Interestingly, none of the tests fail because of that. I suspect the exact set of decompositions here means we could get away with having them Yes, that is, the quickCheck algorithm would still return Maybe or No on a non-normalized sequence, because none of the Yes composites would recombine with each other, see L2/22-157 p. 11. But let’s not try to be too smart.

File: unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java [hunk 1]
```diff
@@ -938,11 +945,260 @@ public static void writeNormalizerTestSuite(String directory, String fileName)
...
+            if (Default.ucd().getCombiningClass(first) == 0) {
+                primaryCompositesByLastNFDCodePoint
+                        .computeIfAbsent(first, key -> new TreeSet<>())
+                        .add(first);
+            }
+            if (Default.ucd().getCombiningClass(second) == 0) {
+                primaryCompositesByFirstNFDCodePoint
+                        .computeIfAbsent(second, key -> new TreeSet<>())
+                        .add(second);
+            }
```
Comment (markusicu) [inline unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java:1039]: Should the ...ByFirst/ByLast maps be keyed by first vs. last code points? It's the opposite right now, and you are mapping each code point to itself.
If it is as intended, then I am confused... maybe a comment in that case?

File: unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java [hunk 2]
```diff
@@ -938,11 +945,260 @@ public static void writeNormalizerTestSuite(String directory, String fileName)
...
+        // canonical decomposition.
+        Set<String> links = new TreeSet<>();
+        for (String decomposition : canonicalDecompositionsOfSingleCodepoints) {
+            int first = decomposition.codePointAt(0);
+            int second;
+            for (int i = UTF16.getCharCount(first);
+                    i < decomposition.length();
+                    first = second, i += UTF16.getCharCount(first)) {
+                second = decomposition.codePointAt(i);
+                links.add(Character.toString(first) + Character.toString(second));
```
Comment (markusicu) [inline unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java:1007]: optional, maybe wrong:

It seems like you could populate the primaryCompositesBy(First,Last)NFDCodePoint maps directly here, without collecting a list of links...??

File: unicodetools/src/main/java/org/unicode/text/UCD/Normalizer.java [hunk 3]
```diff
@@ -250,7 +250,9 @@ public void getCompositionStatus(BitSet leading, BitSet trailing, BitSet resulti
     }
 
     public boolean isTrailing(int cp) {
-        return composition ? data.isTrailing(cp) : false;
+        var buffer = new StringBuffer();
+        data.getRecursiveDecomposition(cp, buffer, compatibility);
+        return composition ? data.isTrailing(buffer.codePointAt(0)) : false;
```
Comment (markusicu) [inline unicodetools/src/main/java/org/unicode/text/UCD/Normalizer.java:255]: nit / minor performance
```suggestion
        if (!composition) { return false; }
        var buffer = new StringBuffer();
        data.getRecursiveDecomposition(cp, buffer, compatibility);
        return data.isTrailing(buffer.codePointAt(0));
```

**Review by markusicu — COMMENTED** [review]
Results look very good -- especially the NFxC_QC values!
I gave up trying to understand 100% of the generator...

File: unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java (hunk 1 above)
Comment (eggrobin) [reply unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java:1039]: The actual building of the maps is above, ll. 975 _sqq_., where a comment indicates that I will cheesily add some more mappings of codepoints to themselves on top of what the identifier actually describes, which indeed happens here. I will try again to look for a cleaner way to write this.

File: unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java (hunk 2 above)
Comment (eggrobin) [reply unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java:1007]: It has been populated already, except for the cheesy self-mappings added lazily below.

**Review by markusicu — DISMISSED** [review]
(partial rs)lgtm

**Comment by macchiati** [comment]
Note, about the following. I think it would help for us to do the following
in ICU to any UTF16 method that has a corresponding Character method.
a) rewrite the method to call the JDK. That way any IDE refactoring will
allow inlining to convert easily to the JDK method.
b) (eventually) deprecate the method

-            for (int i = UTF16.getCharCount(first);
+            for (int i = Character.charCount(first);

File: unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java (hunk 1 above)
Comment (eggrobin) [reply unicodetools/src/main/java/org/unicode/text/UCD/GenerateData.java]: I have rewritten it without that cheesy mutability. PTAL.

**Review by markusicu — APPROVED** [review]

**Pull request merged**